plan/rule_eliminate_projection.go

```diff
@@ -51,9 +51,9 @@ func canProjectionBeEliminatedStrict(p *PhysicalProjection) bool {
 func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
 	dst := replace[string(origin.HashCode(nil))]
 	if dst != nil {
-		colName := origin.ColName
+		colName, retType := origin.ColName, origin.RetType
```

Comment:
why make this change?

Reply:
e.g.
``` sql
tidb> desc select distinct a from t2 ;
+---------------------+----------+------+-------------------------------------------------------------+
| id                  | count    | task | operator info                                               |
+---------------------+----------+------+-------------------------------------------------------------+
| HashAgg_4           | 8000.00  | root | group by:test.t2.a, funcs:firstrow(test.t2.a)               |
| └─TableReader_8     | 10000.00 | root | data:TableScan_7                                            |
|   └─TableScan_7     | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo |
+---------------------+----------+------+-------------------------------------------------------------+
3 rows in set (0.00 sec)
```

There will be a Projection between HashAgg4 and TableReader_8 during logical plan building,
whose schema is `[a (type enum)]`.
We reset the return type of `origin` (which is from HashAgg4's schema) here to make sure the return type of HashAgg_4 is the result of type inferring rather than the return type of Projection.